### PR TITLE
bot efficiency improvements; bug fixes

### PR DESCRIPTION
**Pull request by NickAragua**
This is a companion to my previous PR.

It uses the board clustering mechanism introduced there for a massive performance improvement in bot deployment speed. Additionally, it fixes some weird behaviors related to the long-range pathfinding algorithm generating highly sub-optimal paths and trying to blow up buildings it really didn't need to.

As a bonus, the bot should no longer fire on crippled turrets or crippled infantry in buildings (there were some cases where it still would).

### CODE REVIEW

File: megamek/src/megamek/client/bot/BotClient.java [hunk 1]
```diff
@@ -915,33 +920,30 @@ protected Coords getFirstValidCoords(Entity deployedUnit, List<Coords> possibleD
...
-     * current position.
-     * @param entity
-     * @param board
-     * @return
+     * current position. Returns 0 if this can be accomplished without destroying any terrain, 
+     * -50 if this can be accomplished but terrain must be destroyed,
+     * -100 if this cannot be accomplished at all
      */
-    private boolean hasPathToEdge(Entity entity, IBoard board) {
+    private int hasPathToEdge(Entity entity, IBoard board) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/bot/BotClient.java:929]: Could we rename this to `fitnessOfPathToEdge` or something?

File: megamek/src/megamek/client/bot/BotClient.java [hunk 2]
```diff
@@ -915,33 +920,30 @@ protected Coords getFirstValidCoords(Entity deployedUnit, List<Coords> possibleD
...
-            return true;
+            return 0;
         }
         
-        MovePath mp = getBoardEdgePathFinder(entity).findPathToEdge(entity);
-        return mp != null;
+        CardinalEdge destinationEdge = BoardUtilities.determineOppositeEdge(entity);
+        
+        int noReductionZoneSize = getClusterTracker().getDestinationCoords(entity, destinationEdge, false).size();
+        int reductionZoneSize = getClusterTracker().getDestinationCoords(entity, destinationEdge, true).size();
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/bot/BotClient.java:938]: If this is expensive you could move this into an else block after the `noReductionZoneSize > 0` check.

File: megamek/src/megamek/common/Tank.java [hunk 3]
```diff
@@ -580,59 +580,60 @@ public boolean isLocationProhibited(Coords c, int currElevation) {
 
         boolean hasFlotationHull = hasWorkingMisc(MiscType.F_FLOTATION_HULL);
         boolean isAmphibious = hasWorkingMisc(MiscType.F_FULLY_AMPHIBIOUS);
+        boolean hexHasRoad =  hex.containsTerrain(Terrains.ROAD);
 
         switch (movementMode) {
             case TRACKED:
                 if (!isSuperHeavy()) {
-                    return (hex.terrainLevel(Terrains.WOODS) > 1)
+                    return (hex.terrainLevel(Terrains.WOODS) > 1) && !hexHasRoad
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/Tank.java:588]: These and's and or's are confusing as to who binds to who, so I'm struggling to determine if these are "correct".

File: megamek/src/megamek/common/pathfinder/BoardClusterTracker.java [hunk 4]
```diff
@@ -157,18 +192,18 @@ public void updateMovableAreas(Entity entity) {
...
-                                entity.hasWorkingMisc(MiscType.F_FULLY_AMPHIBIOUS) ||
-                                entity.hasWorkingMisc(MiscType.F_LIMITED_AMPHIBIOUS);
+        MovementType movementType = MovementType.getMovementType(entity);
+        boolean isHovercraft = movementType == MovementType.Hover;
+        boolean isAmphibious = movementType == MovementType.WheeledAmphi ||
+                                movementType == MovementType.TrackedAmphi;
         
         for(int x = 0; x < board.getWidth(); x++) {
-            for(int y = 0; y < board.getHeight(); y++) {
+            for(int y = 0; y < board.getHeight(); y++) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/pathfinder/BoardClusterTracker.java:201]:
```suggestion
            for(int y = 0; y < board.getHeight(); y++) {
```

File: megamek/src/megamek/client/bot/BotClient.java (hunk 2 above)
Comment (NickAragua) [reply megamek/src/megamek/client/bot/BotClient.java:938]: Running time of n hashmap lookups where n is the width/height of the edge, so probably not worth it.

File: megamek/src/megamek/common/Tank.java (hunk 3 above)
Comment (NickAragua) [reply megamek/src/megamek/common/Tank.java:588]: I can put 'em into parens, and write up some comments.

File: megamek/src/megamek/common/MovePath.java [hunk 5]
```diff
@@ -1840,4 +1833,26 @@ public boolean setStrafingStep(Coords pos) {
     public boolean nextForwardStepOffBoard() {
         return !game.getBoard().contains(getFinalCoords().translated(getFinalFacing()));
     }
+    
+    /**
+     * Debugging method that calculates a destruction-aware move path to the destination coordinates
+     */
+    public MovePath calculateDestructionAwarePath(Coords dest) {
+        // code that's useful to test the destruction-aware pathfinder
+        // remove when code review and testing complete
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply megamek/src/megamek/common/MovePath.java:1842]: Guess not anymore.

File: megamek/src/megamek/common/util/BoardUtilities.java [hunk 6]
```diff
@@ -1564,7 +1566,7 @@ private static int normRNG(int factor) {
     }
 
     /**
-     * Figures out the "opposite" edge for the given entity on the entity's game board
+     * Figures out the "closest" edge for the given entity on the entity's game board
      * @param entity Entity to evaluate
      * @return the Board.START_ constant representing the "opposite" edge
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply megamek/src/megamek/common/util/BoardUtilities.java:1571]: Yep.

**Pull request merged**